torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -840,33 +852,156 @@ def test_build_rpc_profiling_key(self):
             self.assertIn("worker0", rpc_profiling_key)
             self.assertIn("worker1", rpc_profiling_key)
 
-    def _profiler_test_with_rpc(self, rpc_exec_mode, func, args, use_record_function=False):
+    def check_profiling_info(self, self_worker_name, dst_worker_name, func, rpc_event, rpc_exec_mode):
+        self.assertTrue(self_worker_name in rpc_event.name)
+        self.assertTrue(dst_worker_name in rpc_event.name)
+        if isinstance(func, torch.jit.ScriptFunction):
+            self.assertTrue(torch.jit._qualified_name(func) in rpc_event.name)
+        else:
+            self.assertTrue(func.__name__ in rpc_event.name)
+        self.assertTrue(rpc_exec_mode.value in rpc_event.name)
+        self.assertEqual(rpc_event.count, 1)
+
+    @dist_init
+    def test_profiler_remote_events_profiled(self):
+        # Tests that we can successfully invoke the profiler on a remote node,
+        # and collect the remote events back in the local profiler.
+        if self.rank != 1:
+            return
+
+        dst_ranks = [rank for rank in range(self.world_size) if rank != self.rank]
+        for dst in dst_ranks:
+            dst_worker = worker_name(dst)
+            with torch.autograd.profiler.profile() as prof:
+                fut = rpc.rpc_async(dst_worker, udf_with_torch_ops, args=())
+                ret = fut.wait()
+
+            events = prof.function_events
+
+            rpc_event = get_function_event(events, RPCExecMode.ASYNC.value)
+            self.check_profiling_info(
+                worker_name(self.rank),
+                dst_worker,
+                udf_with_torch_ops,
+                rpc_event,
+                RPCExecMode.ASYNC,
+            )
+            expected_remote_events = [
+                "mul",
+                "add",
+                "ones",
+                "relu",
+                "sigmoid",
+                "threshold",
+            ]
+            remote_events = {
+                event.name: event
+                for event in events
+                if event.name in expected_remote_events
+            }
+            for expected_remote_event_name in expected_remote_events:
+                self.assertTrue(expected_remote_event_name in remote_events)
+                remote_event = remote_events[expected_remote_event_name]
+                self.assertEqual(remote_event.node_id, dst)
+
+    @dist_init
+    def test_profiler_with_autograd_context(self):
         dst = (self.rank + 1) % self.world_size
-        # only run profiler on rank 1.
         if self.rank == 1:
+            # Case where we can double wrap messages with profiling information and autograd info.
             with torch.autograd.profiler.profile() as prof:
```

Comment:
Is it allowed to flip the order of profiler and dist_autograd contexts?

Reply:
Yes, let me add a UT with this.

Reply:
Added UT with flipped order.